src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServer.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using StreamJsonRpc;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public abstract class LanguageServer<RequestContextType> : ILanguageServer
+{
+    protected readonly ILspLogger _logger;
+
+    protected readonly string _serverKind;
+
+    public bool IsInitialized { get; private set; } = false;
+
+    public bool HasShutdownStarted => _shuttingDown;
+
+    protected LanguageServer(
+        JsonRpc jsonRpc,
+        ILspLogger logger,
+        string serverKind)
+    {
+        throw new NotImplementedException();
+    }
+
+    /// <summary>
+    /// This spins up the LSP and should be called at the bottom of the constructor of the non-abstract implementor.
+    /// </summary>
+    public virtual void Initialize()
+    {
+        GetRequestExecutionQueue();
+        GetRequestDispatcher();
+    }
+
+    protected abstract ILspServices GetLspServices();
+
+    protected virtual IRequestDispatcher<RequestContextType> ConstructDispatcher()
```

Comment:
should these be virtual? or abstract.  what happens if this isn't overridden?

Reply:
Personally, I like that most of these are virtual. I think CLaSP should provide as much as it can by way of default implementations of things, but being virtual stull allows customization.

Reply:
@davidwengier encapsulated my thoughts here perfectly.

If the user doesn't override `ConstructDispatcher` then it will construct the [default `RequestDispatcher` implementation](https://github.com/dotnet/roslyn/pull/62924/files#diff-68827171a9800c8a40f4ccd71a96a93fb5206420213e3532846c7ab0fa0f453fR69-R76).